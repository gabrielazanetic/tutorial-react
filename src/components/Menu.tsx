import { FC, ReactNode } from "react";

interface MenuProps {
    children: ReactNode,
    title?: string
    collapsed?: boolean
}

const Menu: FC<MenuProps> = ({ children, title, collapsed }) => {
    const classes = "menu" + (collapsed ? " collapsed" : "");

    return (
        <>
            {title && <h4>{title}</h4>}
            <ul className={classes}>{children}</ul>
        </>
    );
}

export default Menu;