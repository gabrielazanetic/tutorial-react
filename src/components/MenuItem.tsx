import { FC, ReactNode } from "react";

interface MenuItemProps {
	children: ReactNode | string;
}

const MenuItem: FC<MenuItemProps> = ({ children }) => {
	return <li className="menu__item">{children}</li>;
};

export default MenuItem;
