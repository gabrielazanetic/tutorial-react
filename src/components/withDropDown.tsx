import { ReactNode, useState } from "react";

interface SubmenuProps {
    children: ReactNode,
    title: string | ReactNode,
    collapsed: boolean,
    className?: string
}

const withDropDown = (Menu: any) => {
    return ({ children, title, collapsed = false, className = '' }: SubmenuProps) => {
        const [isCollapsed, setIsCollapsed] = useState(collapsed);
    
        const classes = (isCollapsed ? " collapsed" : "") 
            + (className.length ? " " + className : "");
    
        const openMenu = () => {
            setIsCollapsed(false);
        }
    
        const closeMenu = () => {
            setIsCollapsed(true);
        };
    
        return (
            <div onMouseLeave={closeMenu}>
                {title && <div onMouseEnter={openMenu}>{title}</div>}
                <Menu className={classes}>{children}</Menu>
            </div>
        );
    };
};

export default withDropDown;