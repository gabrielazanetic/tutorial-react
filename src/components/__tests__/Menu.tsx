import renderer from "react-test-renderer";
import MenuItem from "../MenuItem";
import withDropDown from "../withDropDown";
import Menu from "../Menu";
import { MemoryRouter, NavLink } from "react-router-dom";

it("changes class when hovered", () => {
	const SubMenu = withDropDown(Menu);

	const menuItem = renderer.create(
		<MemoryRouter initialEntries={["/"]}>
			<SubMenu title={<NavLink to="/">Home</NavLink>} collapsed={true}>
				<MenuItem>Step 1</MenuItem>
				<MenuItem>Step 2</MenuItem>
				<MenuItem>Step 3</MenuItem>
				<MenuItem>Step 4</MenuItem>
				<MenuItem>Step 5</MenuItem>
			</SubMenu>
		</MemoryRouter>
	);

	let tree: any = menuItem.toJSON();

	expect(tree).toMatchSnapshot();

	renderer.act(() => {
		tree.children[0].props.onMouseEnter();
	});

	tree = menuItem.toJSON();
	expect(tree).toMatchSnapshot();

	renderer.act(() => {
		tree.props.onMouseLeave();
	});

	tree = menuItem.toJSON();
	expect(tree).toMatchSnapshot();
});
