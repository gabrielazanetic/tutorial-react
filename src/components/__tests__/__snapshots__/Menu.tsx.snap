// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes class when hovered 1`] = `
<div
  onMouseLeave={[Function]}
>
  <div
    onMouseEnter={[Function]}
  >
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
  </div>
  <ul
    className="menu  collapsed"
  >
    <li
      className="menu__item"
    >
      Step 1
    </li>
    <li
      className="menu__item"
    >
      Step 2
    </li>
    <li
      className="menu__item"
    >
      Step 3
    </li>
    <li
      className="menu__item"
    >
      Step 4
    </li>
    <li
      className="menu__item"
    >
      Step 5
    </li>
  </ul>
</div>
`;

exports[`changes class when hovered 2`] = `
<div
  onMouseLeave={[Function]}
>
  <div
    onMouseEnter={[Function]}
  >
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
  </div>
  <ul
    className="menu"
  >
    <li
      className="menu__item"
    >
      Step 1
    </li>
    <li
      className="menu__item"
    >
      Step 2
    </li>
    <li
      className="menu__item"
    >
      Step 3
    </li>
    <li
      className="menu__item"
    >
      Step 4
    </li>
    <li
      className="menu__item"
    >
      Step 5
    </li>
  </ul>
</div>
`;

exports[`changes class when hovered 3`] = `
<div
  onMouseLeave={[Function]}
>
  <div
    onMouseEnter={[Function]}
  >
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
  </div>
  <ul
    className="menu  collapsed"
  >
    <li
      className="menu__item"
    >
      Step 1
    </li>
    <li
      className="menu__item"
    >
      Step 2
    </li>
    <li
      className="menu__item"
    >
      Step 3
    </li>
    <li
      className="menu__item"
    >
      Step 4
    </li>
    <li
      className="menu__item"
    >
      Step 5
    </li>
  </ul>
</div>
`;
